CIPs/CIP-69/CIP-69.md

```diff
@@ -0,0 +1,108 @@
+
+
+## Specification
+
+First lets define the basic IPLD data structures used.
+```ipldsch
+type MerkleNode [Link]
+
+type TreeMetaData struct {
+  numEntries Int
+  bloomType String
+  bloomFilter {String:Any}
+}
+```
+Here the *MerkleNode* is simply an array of CIDs, which is used as the main building block for the merkle tree. The *TreeMetaData* contains some metadata about the tree, namely the number of entries and a bloom filter. The `bloomType` is used to describe which specific type of bloom filter is being used and is there for upgradeability purposes.
+
+### Tree structure
+The structure of the merkle tree is very straight forward. Starting from the bottom each individual update is paired with another update and a *MerkleNode* is created which has the leftmost update in position `0`, and the other update at position `1`. *MerkleNodes* are then created on each level until we reach the root. The root *MerkleNode* has an additional entry at position `2` which simply points to a *TreeMetaData* object. In the figure below we see a simple tree with four leafs.
+![merkletree(1)](https://user-images.githubusercontent.com/3909429/97851426-a83bf900-1cf5-11eb-9d9f-e7b568138850.png)
+In this example an *IPLD Path* can be used to reach any of the leafs. For example, to reach *Update 3* we would use the path `<root-cid>/1/0`. Similarly we can always get the *TreeMetaData* using the path `<root-cid>/2`.
+
+The merkle tree should be constructed as a balanced tree. See the [implementation][https://github.com/ceramicnetwork/ceramic-anchor-service/blob/4615c814ed98bb17cba8902480d8f22bde5397a1/src/merkle/merkle-tree.ts#L46-L65] for reference.
+
+
+### Leaf sorting
+Before creating the tree the leafs (updates) should be sorted. The sorting should be based on a few different properties of the Ceramic document that is being updated and the update record itself:
+
+1. `topic` - sort by the topic in the document metadata
+2. `schema` - if multiple documents have the same topic, sort by the *schema*
+3. `DID` - if multiple documents have the same topic and schema, extract and sorty by the DID from the `kid` in the signed record
+4. `DocID` - finally sort by the DocID
+
+
+### Bloom filter
+The bloom filter should include the following data for each document that is being updated:
+
+* The `topic`, prepend each tag string with `topic-`
+
+* The first 5 `tags`, prepend each tag string with `tag-`
+* The `schema`, prepend the schema DocID string with `schema-`
+* All DID strings in the `controllers` array, prepend each DID with `did-`
+* The DocID string of the document
+
+The bloom filter is created using the javascript [bloom-filter](https://github.com/Callidon/bloom-filters) library. Specifically using the *Classic Bloom Filter*. An example for how to create the filter can be observed below.
+
+```js
+const { BloomFilter } = require('bloom-filters')
+const filterEntries = [...] // An array of strings
+const errorRate = 0.04 // 4 % error rate
```

Comment:
Why .04?  Was that chosen arbitrarily or based on some calculations about the size of the resulting filter?

Reply:
Kind of arbitrarily. Just took it out of the readme of the `bloom-filters` package.

Reply:
Ah I see.  I guess I had been thinking this was supposed to be an example of how someone might consume the information stored in the `bloomFilter` field of the `TreeMetadata` to reconstitute the bloom filter and use it, but I guess this is more like example code of how **we** might build the bloom filter before persisting it in the TreeMetadata.  In that case, is this example even really relevant/useful to include in this document?

Reply:
i don't think it hurts to include?

Reply:
Only risk is confusion/distraction from the main focus of how to use/consume this structure.  Probably not a big deal either way though